### PR TITLE
Refactor timezones functions out of tslib

### PR DESCRIPTION
**Pull request by jbrockmendel**
`_libs.tslib` is over 5k lines and is imported by a bunch of other modules including `_libs.lib`.  It looks like it was pasted together from a bunch of older files.  There are a handful of areas where significant chunks can be refactored out in complexity-reducing (and testability-increasing) ways.  This is the first one: timezones.  (Next up: parsing (and gathering parsing code dispersed across pandas))

timezones.pyx has no other dependencies within pandas, helping to de-tangle some of the `_libs` modules

Code in timezones is used in both `_libs.tslib` and `_libs.period`, and a bit in `_libs.index`.

This is one of several steps in making _libs.period not need to import from non-cython code and
ideally not need to import `tslib` (though `NaT` makes that tough).  See existing comments
in `_libs.__init__` on why this is desireable.

This is the first of several independent pieces to be split off of `tslib`.

There are also several notes on functions that appear to be unused and may be ready for removal.

Removes `datetime_helper` dependency from most of `_libs`, as it is somehow slower than a plain cython version.  In cases where C can be replaced by idiomatic cython without hurting performance, I'm calling that a win.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/period.pyx [hunk 1]
```diff
@@ -10,19 +10,19 @@ from numpy cimport (int8_t, int32_t, int64_t, import_array, ndarray,
...
 import numpy as np
 
-cdef extern from "datetime_helper.h":
-    double total_seconds(object)
-
 from libc.stdlib cimport free
 
 from pandas import compat
 from pandas.compat import PY2
+from pandas.core.dtypes.generic import ABCDateOffset
```
Comment (jbrockmendel) [inline pandas/_libs/period.pyx]: An upcoming PR will make a cython version of `dtypes.generic`.  The `isinstance` checks are about 2x faster that way.  It's a small difference, but nice to avoid calling into python for these things.

File: pandas/_libs/period.pyx [hunk 2]
```diff
@@ -771,9 +779,7 @@ cdef class _Period(object):
 
     def __add__(self, other):
         if isinstance(self, Period):
-            if isinstance(other, (timedelta, np.timedelta64,
-                                  offsets.Tick, offsets.DateOffset,
-                                  Timedelta)):
+            if isinstance(other, (timedelta, np.timedelta64, ABCDateOffset)):
```
Comment (jbrockmendel) [inline pandas/_libs/period.pyx]: `Timedelta` subclasses `timedelta` and `Tick` subclasses `DateOffset`, so we can remove them from these checks.

File: pandas/_libs/tslib.pyx [hunk 3]
```diff
@@ -3960,7 +3889,7 @@ for _maybe_method_name in dir(NaTType):
             def f(*args, **kwargs):
                 raise ValueError("NaTType does not support " + func_name)
             f.__name__ = func_name
-            f.__doc__ = _get_docstring(_method_name)
+            f.__doc__ = _get_docstring(func_name)
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx]: This is an unrelated bug.  This block of code exists to make `NaT` raise `ValueError` for each of a bunch of `datetime` methods.  This is supposed to attach the original docstring to the new function `f`.  The typo `_method_name` instead of `func_name` means that a bunch of the docstrings are currently wrong.

**Comment by gfyoung** [comment]
> Removes datetime_helper dependency from most of _libs, as it is somehow slower than a plain cython version.

Do you have any (empirical) timing information to support this statement?

**Comment by jbrockmendel** [comment]
> Do you have any (empirical) timing information to support this statement?

Brief discussion of profiling results [here](https://github.com/pandas-dev/pandas/issues/17234#issuecomment-322932822).

**Comment by jorisvandenbossche** [comment]
Regarding the `total_seonds` in `datetime_helper.h`, a comment mentions there it is for 2.6 compat, so it can maybe be removed altogether? (without a cython replacement)

**Comment by jbrockmendel** [comment]
> [...] so it can maybe be removed altogether?

That would be excellent.  If there's a concensus, I'll amend the PR.

**Comment by jbrockmendel** [comment]
> but since we dropped 2.6 support long ago, we can just use the total_seconds method instead of our custom function.

@jorisvandenbossche I'll give that a try and see how it goes.  Is it possible/likely that this would also explain why _libs/src/datetime/np_datetime.c duplicates a bunch of stuff from numpy 1.7?  i.e. at a time when pre-1.7 versions of numpy were supported?

**Comment by jbrockmendel** [comment]
Getting rid of `total_seconds` (and `tot_seconds`) in all of the pyx files works.  There is one call from objToJSON.c that someone else should take a look at.

File: pandas/_libs/period.pyx (hunk 1 above)
Comment (jreback) [inline pandas/_libs/period.pyx]: let's keep perf changes separate. I would much prefer PR's that don't mix multiple changes all at once.

Comment (jbrockmendel) [inline pandas/_libs/period.pyx]: OK.  I just pushed a commit that reverted from `ABCDateOffset` back to `offsets.DateOffset`.

File: pandas/_libs/tslib.pyx (hunk 3 above)
Comment (jreback) [inline pandas/_libs/tslib.pyx]: then pls don't change here. rather in another PR.

File: pandas/_libs/tslibs/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
```
Comment (jreback) [inline pandas/_libs/tslibs/__init__.py]: why are you adding this? this is NOT a package by definition.

File: pandas/_libs/tslibs/timezones.pyx [hunk 5]
```diff
@@ -0,0 +1,316 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# cython: profile=False
+
+try: string_types = basestring
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: what the heck is this?

this is the point of pandas.compat

File: pandas/_libs/tslibs/timezones.pyx [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+# dateutil compat
+from dateutil.tz import (tzoffset,
+                         tzlocal as _dateutil_tzlocal,
+                         tzfile as _dateutil_tzfile,
+                         tzutc as _dateutil_tzutc,
+                         tzstr as _dateutil_tzstr)
+
+import sys
+if sys.platform == 'win32' or sys.platform == 'cygwin':
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: and why are you not using the idiomatic

File: pandas/_libs/tslibs/timezones.pyx [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+cdef inline bint is_integer_object(object obj):
+    # Ported from util (which gets it from numpy_helper) to avoid
+    # direct dependency
+    return (not PyBool_Check(obj)) and PyArray_IsIntegerScalar(obj)
+
+
+cdef int64_t NPY_NAT = np.datetime64('nat').astype(np.int64)
+
+
+# TODO: Does this belong somewhere else?
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: why is this here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls pls pls limit the changes to the minimal set that are required. mixing changes makes this much harder to review.

File: pandas/_libs/tslibs/timezones.pyx (hunk 5 above)
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timezones.pyx]: Trying to avoid non-cython imports. Will change.

File: pandas/_libs/tslib.pyx (hunk 3 above)
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx]: Will change.  Apologies.

File: pandas/_libs/tslibs/__init__.py (hunk 4 above)
Comment (jbrockmendel) [inline pandas/_libs/tslibs/__init__.py]: So that it can be imported python-style.  Or if you're referring to the shebang that's force of habit.

File: pandas/_libs/tslibs/timezones.pyx (hunk 6 above)
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timezones.pyx]: That was copied verbatim from pd.compat, so I'm not sure what idiom it misses.  But as above the idea was to avoid a non-cython import.  Will change.

File: pandas/_libs/tslibs/timezones.pyx (hunk 7 above)
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timezones.pyx]: This PR predates #17289.  Specifying C dependencies in setup.py is a PITA and I wanted this module to be portable.  So I copied `total_seconds` instead of getting it from src/datetime_helper.h.

Regardless of whether #17289 is merged, this function can be removed.  Coming right up.

**Comment by jbrockmendel** [comment]
Gonna push on this a little bit.  While #17363 and #17342 are "just" about making code more self-contained and maintainable, this PR is a blocker for some more performance-relevant fixes.

In particular, the end goal is to move `tseries.offsets.DateOffset` and `tseries.frequencies.to_offset` into cython, avoid the runtime import of `to_offset`, and make `DateOffset` immutable.  Being able to import timezone functions without importing all of `tslib` makes it possible to move large chunks of `frequencies` and `offsets` into cython.

File: pandas/_libs/tslib.pyx [hunk 8]
```diff
@@ -4081,6 +4013,7 @@ def pydt_to_i8(object pydt):
     return ts.value
 
 
+# TODO: Never used?
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: I think you can take this out (this function)

File: pandas/_libs/tslibs/__init__.py (hunk 4 above)
Comment (jreback) [reply pandas/_libs/tslibs/__init__.py]: yes pls don't add shebangs anywhere.

Comment (jreback) [reply pandas/_libs/tslibs/__init__.py]: generally you should be adding ``__init__.pyx`` for cython packages (IIRC you can then still import)

File: pandas/_libs/tslibs/timezones.pyx [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+import pytz
+UTC = pytz.utc
+
+
+from cpython cimport PyBool_Check
+cdef extern from "numpy/ndarrayobject.h":
+    bint PyArray_IsIntegerScalar(object)
+
+cdef inline bint is_integer_object(object obj):
+    # Ported from util (which gets it from numpy_helper) to avoid
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: not a useful comment. Why are you not simply importing from numpy_helper?

File: pandas/_libs/tslibs/timezones.pyx [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+
+
+#----------------------------------------------------------------------
+# time zone conversion helpers
+
+def get_timezone(tz):
+    return _get_zone(tz)
+
+
+cdef bint _is_utc(object tz):
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx]: I think these can all be inline (though if they give warning then prob not)

**Review by jreback — CHANGES_REQUESTED** [review]
I want to see an asv run of all timezone & time realted benchmarks. (IOW show the diffs)

File: pandas/_libs/tslibs/__init__.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/__init__.py]: Will change.

File: pandas/_libs/tslibs/timezones.pyx (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timezones.pyx]: Will change.

The thought was that the marginal flexibility of going from 1 intra-pandas dependency to 0 was big enough to merit the redundant one-liner.  Plus avoids having to tinker with setup.py and debug the inevitable segfaults that go with it.

File: pandas/_libs/tslibs/timezones.pyx (hunk 10 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timezones.pyx]: will change.

File: pandas/_libs/tslibs/__init__.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/__init__.py]: > generally you should be adding __init__.pyx for cython packages (IIRC you can then still import)
 
Just renaming led to `ImportError`.  Does something need to go in setup.py for this to work?

**Comment by jbrockmendel** [comment]
> I want to see an asv run of all timezone & time realted benchmarks. (IOW show the diffs)

asv is complaining about build errors, so I'll put together a handful of measurements just using %timeit.  There isn't a Timestamp/timezone-specific benchmark file,  though I think there is one added in #17331

File: pandas/_libs/tslibs/timezones.pyx (hunk 9 above)
Comment (jreback) [reply pandas/_libs/tslibs/timezones.pyx]: repeating low-level code is much worse

**Comment by jreback** [comment]
> asv is complaining about build errors, so I'll put together a handful of measurements just using %timeit. There isn't a Timestamp/timezone-specific benchmark file, though I think there is one added in #17331

pls don't. we use asv for a reason to avoid ad-hoc things like this.

**Comment by jbrockmendel** [comment]
> pls don't. we use asv for a reason to avoid ad-hoc things like this.

OK... what's the difference between this and the other times you've told me to use %timeit?

```
asv continuous -f 1.1 -E virtualenv HEAD -b timeseries.py
· Creating environments
· Discovering benchmarks
·· Uninstalling from virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytest-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt
·· Installing into virtualenv-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytest-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt..
·· Error running ~/pd/tslibs-timezones/pandas/asv_bench/env/dbb8113eac34359a9ef26b93e54a1d03/bin/python ~/.local/lib/python2.7/site-packages/asv/benchmark.py discover ~/pd/tslibs-timezones/pandas/asv_bench/benchmarks /tmp/tmpUDGky2
             STDOUT -------->
             
             STDERR -------->
             Traceback (most recent call last):
               File "~/.local/lib/python2.7/site-packages/asv/benchmark.py", line 867, in <module>
                 commands[mode](args)
               File "~/.local/lib/python2.7/site-packages/asv/benchmark.py", line 812, in main_discover
                 list_benchmarks(benchmark_dir, fp)
               File "~/.local/lib/python2.7/site-packages/asv/benchmark.py", line 797, in list_benchmarks
                 for benchmark in disc_benchmarks(root):
               File "~/.local/lib/python2.7/site-packages/asv/benchmark.py", line 696, in disc_benchmarks
                 for module in disc_files(root, os.path.basename(root) + '.'):
               File "~/.local/lib/python2.7/site-packages/asv/benchmark.py", line 648, in disc_files
                 module = import_module(package + filename)
               File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
                 __import__(name)
               File "~/pd/tslibs-timezones/pandas/asv_bench/benchmarks/index_object.py", line 1, in <module>
                 from .pandas_vb_common import *
               File "~/pd/tslibs-timezones/pandas/asv_bench/benchmarks/pandas_vb_common.py", line 1, in <module>
                 from pandas import *
               File "~/pd/tslibs-timezones/pandas/asv_bench/env/dbb8113eac34359a9ef26b93e54a1d03/local/lib/python2.7/site-packages/pandas/__init__.py", line 35, in <module>
                 "the C extensions first.".format(module))
             ImportError: C extension: No module named tslibs.timezones not built. If you want to import pandas from the source directory, you may need to run 'python setup.py build_ext --inplace --force' to build the C extensions first.
```

Same deal with py3.5 except it's complaining about `period.pquarter`

**Comment by jreback** [comment]
this ian changing way more not-easy-to-inspect code differences

if you are going to change significant code then get asv to work
it should be quite straightforward 

i have never used this with virtual env can't say if it actually works properly (use conda)

**Comment by jbrockmendel** [comment]
The same error is occuring with conda.  The same "ImportError: C extension: No module named tslibs.timezones not built." is occuring even if I just run `asv continuous -f 1.1 master -b ^timeseries` where there _should be no tslibs.timezones`.  It's 100° out and I'm not inclined to spend any more time on this.

I'll make a series of PRs that move these functions over one at a time.

**Comment by jreback** [comment]
closing this as its replaced by other PR's. let's close older / extraneous PR's.
